src/Compilers/Core/Portable/SourceGeneration/IncrementalGeneratorInitializationContext_ForAttributeWithMetadataName.cs

```diff
@@ -0,0 +1,167 @@
+
+    /// <summary>
+    /// Semantic model for the file that <see cref="AttributeTarget"/> is contained within.
+    /// </summary>
+    public SemanticModel SemanticModel { get; }
+
+    /// <summary>
+    /// <see cref="AttributeData"/>s for any matching attributes on <see cref="AttributeTarget"/>.  Always non-empty.
+    /// </summary>
+    public ImmutableArray<AttributeData> Attributes { get; }
+
+    internal GeneratorAttributeSyntaxContext(
+        SyntaxNode attribueTarget,
+        SemanticModel semanticModel,
+        ImmutableArray<AttributeData> attributes)
+    {
+        AttributeTarget = attribueTarget;
+        SemanticModel = semanticModel;
+        Attributes = attributes;
+    }
+}
+
+public partial struct IncrementalGeneratorInitializationContext
+{
+    private static readonly char[] s_nestedTypeNameSeparators = new char[] { '+' };
+    private static readonly SymbolDisplayFormat s_metadataDisplayFormat =
+        SymbolDisplayFormat.QualifiedNameArityFormat.AddCompilerInternalOptions(SymbolDisplayCompilerInternalOptions.UsePlusForNestedTypes);
+
+    /// <summary>
+    /// Creates an <see cref="IncrementalValuesProvider{T}"/> that can provide a transform over all <see
+    /// cref="SyntaxNode"/>s if that node has an attribute on it that binds to a <see cref="INamedTypeSymbol"/> with the
+    /// same fully-qualified metadata as the provided <paramref name="fullyQualifiedMetadataName"/>. <paramref
+    /// name="fullyQualifiedMetadataName"/> should be the fully-qualified, metadata name of the attribute, including the
+    /// <c>Attribute</c> suffix.  For example <c>"System.CLSCompliantAttribute</c> for <see
+    /// cref="System.CLSCompliantAttribute"/>.
+    /// </summary>
+    /// <param name="predicate">A function that determines if the given <see cref="SyntaxNode"/> attribute target (<see
+    /// cref="GeneratorAttributeSyntaxContext.AttributeTarget"/>) should be transformed.  Nodes that do not pass this
+    /// predicate will not have their attributes looked at at all.</param>
+    /// <param name="transform">A function that performs the transform. This will only be passed nodes that return <see
+    /// langword="true"/> for <paramref name="predicate"/> and which have a matchin <see cref="AttributeData"/> whose
+    /// <see cref="AttributeData.AttributeClass"/> has the same fully qualified, metadata name as <paramref
+    /// name="fullyQualifiedMetadataName"/>.</param>
+    internal IncrementalValuesProvider<T> ForAttributeWithMetadataName<T>(
+        string fullyQualifiedMetadataName,
+        Func<SyntaxNode, CancellationToken, bool> predicate,
+        Func<GeneratorAttributeSyntaxContext, CancellationToken, T> transform)
+    {
+        var metadataName = fullyQualifiedMetadataName.Contains('+')
+            ? MetadataTypeName.FromFullName(fullyQualifiedMetadataName.Split(s_nestedTypeNameSeparators).Last())
+            : MetadataTypeName.FromFullName(fullyQualifiedMetadataName);
```

Comment:
I'm thinking about the `file class` feature that is in progress. That will make it possible to define attributes that have mangled names in metadata. Will those be properly handled by this code? 

Essentially consider what happens when a customer writes the following:

```c#
[AttributeTargets(AttributeTargets.All)]
file class TrickyAttribute { } 

[Tricky]
class C { }
```

Believe this should be a valid scenario with `file class` but will let @RikkiGibson confirm

In practice the name `TrickyAttribute` can never be used in metadata, it can only occur source to source. But the API here is specific about using metadata names which is a bit weird with `file class`.

Reply:
> I'm thinking about the file class feature that is in progress. That will make it possible to define attributes that have mangled names in metadata. Will those be properly handled by this code?

It's not likely to.  What is your name mangling strategy here, and how are you updating GetTypeByMetadataName to support that?  @RikkiGibson ?

Reply:
> Believe this should be a valid scenario with `file class` but will let @RikkiGibson confirm

This is valid--just like how attributes are permitted to be less accessible than the member they're applied to, it's also permitted to apply a `file` attribute to a non-`file` member.

> all should still work as long as the "fully qualified metadata name" at least contains 'Tricky' in it in an easy to grab out fashion

We will do this. The mangling format is not set in stone, but we expect it to be something like `<ContainingFileName>F0__OriginalTypeName`.

> _And_ if the compiler provides a sensible way to go from an INamedTypeSymbol to the fully-qualified-metadata-name for it.

`INamedTypeSymbol.MetadataName` on a file type (from source) will give the mangled name which will be emitted for the type.

I've added a note in #60819 to test GetTypeByMetadataName with file types. I don't anticipate modifying GetTypeByMetadataName behavior directly, only changing what MetadataName is produced for certain symbols.